lib/Onyx.js

```diff
@@ -1047,55 +1047,46 @@ function initializeWithDefaultKeyStates() {
 function clear(keysToPreserve = []) {
     return getAllKeys()
         .then((keys) => {
-            const keyValuesToReset = [];
-            const defaultKeys = _.keys(defaultKeyStates);
-
-            // Get all the values for the keys that need to be preserved. These key/value pairs will be set
-            // in Onyx after the database is cleared().
-            const keyValuesToPreserve = _.map(keysToPreserve, key => [key, cache.getValue(key)]);
+            const keysToBeClearedFromStorage = [];
+            const keyValuesToResetAsCollection = {};
+            const keyValuesToResetIndividually = {};
+            const defaultKeys = _.keys(_.omit(defaultKeyStates, ...keysToPreserve));
```

Comment:
Why do we need the `keysToReset` here? It seems to me like this logic should be covered in the main loop right?

Reply:
Also, if this _is_ needed for some reason I'm not seeing, there's an unnecessary spread:

```suggestion
            const defaultKeys = _.keys(_.omit(defaultKeyStates, keysToPreserve));
```

Reply:
@roryabraham Thanks for your comment. I think we should notify subscribers when data has been cleared as I can see this test on `onyxTest.js`. After checking the old logic I see we push the `defaultKeyValuePairs` to `keyValuesToReset`, that why I decide to have `keysToReset`. You can check the test https://github.com/Expensify/react-native-onyx/blob/4887af4b350de5febe3084e6d13ac07d8f617021/tests/unit/onyxTest.js#L95 
the keys = [ 'test' ] and defaultKeyStates = { otherTest: 42 }, if we don't push `otherTest` to keys, we'll miss this value.
 
And yes, you're right, having spread is unnecessary. I updated my PR to remove the spread.   
https://github.com/Expensify/react-native-onyx/blob/4887af4b350de5febe3084e6d13ac07d8f617021/lib/Onyx.js#L1070-L1072

Reply:
Thanks for explaining. I've spent some time testing, and I still don't think this code is needed. I modified the `onyxTest` a bit to test that default keys are reset:

```diff
diff --git a/tests/unit/onyxTest.js b/tests/unit/onyxTest.js
index 50b8608..e9a955e 100644
--- a/tests/unit/onyxTest.js
+++ b/tests/unit/onyxTest.js
@@ -112,9 +112,10 @@ describe('Onyx', () => {
 
         return waitForPromisesToResolve()
             .then(() => Onyx.set(ONYX_KEYS.TEST_KEY, 'test'))
+            .then(() => Onyx.set(ONYX_KEYS.OTHER_TEST, 420))
             .then(() => {
                 expect(testKeyValue).toBe('test');
-                expect(otherTestValue).toBe(42);
+                expect(otherTestValue).toBe(420);
                 return Onyx.clear().then(waitForPromisesToResolve);
             })
             .then(() => {
```

Then I reset the code to how it was before:

```diff
diff --git a/lib/Onyx.js b/lib/Onyx.js
index 910b759..2667754 100644
--- a/lib/Onyx.js
+++ b/lib/Onyx.js
@@ -1050,16 +1050,13 @@ function clear(keysToPreserve = []) {
             const keysToBeClearedFromStorage = [];
             const keyValuesToResetAsCollection = {};
             const keyValuesToResetIndividually = {};
-            const defaultKeys = _.keys(_.omit(defaultKeyStates, keysToPreserve));
-            const keysToReset = _.difference(keys, defaultKeys);
-            keysToReset.push(...defaultKeys);
 
             // The only keys that should not be cleared are:
             // 1. Anything specifically passed in keysToPreserve (because some keys like language preferences, offline
             //      status, or activeClients need to remain in Onyx even when signed out)
             // 2. Any keys with a default state (because they need to remain in Onyx as their default, and setting them
             //      to null would cause unknown behavior)
-            _.each(keysToReset, (key) => {
+            _.each(keys, (key) => {
                 const isKeyToPreserve = _.contains(keysToPreserve, key);
                 const isDefaultKey = _.has(defaultKeyStates, key);

```

And everything worked as expected. Since this is simpler, let's change it back.